### ISSUE TITLE
Aligning consecutive declarations in enum blocks

### ISSUE THREAD

**Issue by Julien-Elie** [llvm/llvm-project#52983]
Could it be possible to have an option to align declarations in enum blocks?

For instance:

```
enum nntp_code {
    NNTP_INFO_HELP         = 100,
    NNTP_INFO_CAPABILITIES = 101,
    NNTP_INFO_DATE         = 111,
    NNTP_OK_BANNER_POST    = 200,
    NNTP_OK_BANNER_NOPOST  = 201,
    NNTP_OK_QUIT           = 205,
    NNTP_OK_COMPRESS       = 206,
    NNTP_OK_GROUP          = 211,
    NNTP_OK_LIST           = 215,
    NNTP_OK_ARTICLE        = 220,
    NNTP_OK_HEAD           = 221,
    NNTP_OK_BODY           = 222,
}
```

**Comment by mydeveloperday** [comment]
Can you explain why this isn't just `AlignConsecutiveAssignments: Consecutive`

which gives

```
enum nntp_code {
  NNTP_INFO_HELP         = 100,
  NNTP_INFO_CAPABILITIES = 101,
  NNTP_INFO_DATE         = 111,
  NNTP_OK_BANNER_POST    = 200,
  NNTP_OK_BANNER_NOPOST  = 201,
  NNTP_OK_QUIT           = 205,
  NNTP_OK_COMPRESS       = 206,
  NNTP_OK_GROUP          = 211,
  NNTP_OK_LIST           = 215,
  NNTP_OK_ARTICLE        = 220,
  NNTP_OK_HEAD           = 221,
  NNTP_OK_BODY           = 222,
}
```

**Comment by Julien-Elie** [comment]
I agree `AlignConsecutiveAssignments` does the job in `enum` but I would just like to align assignments in `enum`, not everywhere in the source code.

**Comment by mydeveloperday** [comment]
This is doable but would require careful backwards compatibility considerations

**Comment by gapisback** [comment]
Hello, Gentlepeople, I just started looking at LLVM git repo and this set of open issues.

This one to do with `clang-format`-ting is interesting to me as it's a non-core, non-compilation code change.
I am trying to teach myself compiler development, so would be interesting to work on this kind of outside the actual `clang` change.

I'm aware of the backwards-compatibility issue(s) and will take a crack at specifying the interface and behaviour once this issue is assigned to me.

Seems like there is some re-thinking on whether this is a 'new issue' or 'good first issue'; attn. @Endilll -- is this really that difficult? 

[ @Julien-Elie - do you have ownership to assign? ] Will wait for the assigment to happen  before engaging the people who commented on this issue with next-steps and a proposal.

**Comment by Julien-Elie** [comment]
Thanks for having a look at this ticket (as well as other tickets).  Greatly appreciated to go on improving clang-format.

**Comment by gapisback** [comment]
Thank you @Endilll for the quick assignment. I'll start looking at it, @Julien-Elie 

This work-item affects clang-formatting rules and the output of formatted code. There is a bit of backward compatibility too. As this is my very 1st foray into changes in this repo, just asking what the process is here. A specific question:

In past such experiences, elsewhere, when interfaces & output/functionality is involved, I prefer to air out an initial 'mini-spec' describing the new / changed interface and how the affected output will be. Some code-owners sign-off on the interface 'mini-spec', before I turn out the code changes and submit a PR. 

Even though this issue is a small'ish item, I'd like to follow that process as it avoids code rework due to interface changes.

Is this team equipped for such a round-trip pre-review of interfaces? If so, who would I get sign-off from?
(I'll find the last author(s) who worked clang-format tool by `git blame`, but thought I'd ask here if anyone has a quick suggestion of names to run the proposal by.)

**Comment by Julien-Elie** [comment]
@gapisback, did you finally manage to have a look at it?

**Comment by Julien-Elie** [comment]
@mydeveloperday About your suggestion of using `AlignConsecutiveAssignments` (though it applies to all assignments, and not only `enum`), I'm wondering whether it could not be useful to allow specific formatting in some parts of the code.

I explain: instead of totally disabling the reformatting with `/* clang-format off */`, couldn't instead something like `/* clang-format +"AlignConsecutiveAssignments: AcrossEmptyLinesAndComments" */` be used, and when I no longer want it `/* clang-format -"AlignConsecutiveAssignments: AcrossEmptyLinesAndComments" */`?
I don't know whether such a change of style is technically doable within a give file (it may be hard to implement) and if it had already been discussed before.

**Comment by jm4R** [comment]
This idea is a rocket!

**Comment by parmi93** [comment]
I have the same problem too.

What I would like is an option to only align enums and assignments to struct members.

```c
typedef struct
{
    uint8_t version;
    mutex_h mutex;
    bool no_store;
    nvs_inf_h inf_handle;
} nvs_t;

// I like this alignment!
enum nntp_code {
    NNTP_INFO_HELP         = 100,
    NNTP_INFO_CAPABILITIES = 101,
    NNTP_INFO_DATE         = 111
}

void foo(nvs_t **store)
{
    // I also like this alignment!
    (*store)->version    = 1;
    (*store)->mutex      = NULL;
    (*store)->no_store   = true;
    (*store)->inf_handle = NULL;
    
    // I don't want this alilgnment!
    int a                       = 10
    char *print_message_pointer = "Hello World!"
    bool convert                = false;
    int arr[10]                 = {1, 2, 3, 4, 5, 6, 7, 8, 9};
}
```

More generally I would like consecutive assignments to the same object or members of the same object to be aligned.

Is there a way to achieve this?

**Comment by gapisback** [comment]
HI, @Julien-Elie , @mydeveloperday @jm4R @parmi93 -- Folks, Good morning!

Am sorry, but I have to unassign myself from this issue. I had picked it up few months ago when I was in between jobs and excited about learning LLVM / compiler stuff. My excitement has not waned but I am not on a new assignment ... and it's impossible for me to find time to do this item.

Sorry, Julien-Elie -- for letting you down ...

**Comment by Sukumarsawant** [comment]
I think this has been completed by https://github.com/llvm/llvm-project/pull/194154
